File: CPP/Mobile Snippets/Notifications/SendCancelNotification/uMain.cpp
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable,
// as such term is defined thereunder. Your use of this software constitutes
// your acknowledgement of your agreement to the foregoing software license
// and support agreement.
//---------------------------------------------------------------------------

#include <fmx.h>
#include <memory>
#pragma hdrstop

#include "uMain.h"
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"

TNotificationsForm *NotificationsForm;

// ---------------------------------------------------------------------------
__fastcall TNotificationsForm::TNotificationsForm(TComponent *Owner) : TForm(Owner) {
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::NotificationCPermissionRequestResult(TObject *Sender, const bool AIsGranted)
{
	if (AIsGranted && PendingAction)
	{
		PendingAction->Execute();
	}

	PendingAction = nullptr;
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::NotificationCReceiveLocalNotification(TObject *Sender, TNotification *ANotification)
{
	Memo1->Lines->Add(ANotification->AlertBody);
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::ActionListExecute(TBasicAction *Action, bool &Handled)
{
	if (NotificationC->AuthorizationStatus() != TAuthorizationStatus::Authorized)
	{
		Handled = true;
		PendingAction = Action;

		NotificationC->RequestPermission();
	}
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::ActionSendScheduledNotificationExecute(TObject *Sender)
{
	std::unique_ptr<TNotification> Notification(NotificationC->CreateNotification());
	Notification->Name = "MyNotification";
	Notification->AlertBody = "C++Builder for Mobile is here!";
	/* Fired in 10 second */
	Notification->FireDate = Now() + EncodeTime(0, 0, 10, 0);

	/* Send notification in Notification Center */
	NotificationC->ScheduleNotification(Notification.get());
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::ActionSendNotificationImmediatelyExecute(TObject *Sender)
{
	std::unique_ptr<TNotification> Notification(NotificationC->CreateNotification());
	Notification->AlertBody = "C++Builder for Mobile is here!";
	Notification->FireDate = Now();

	// Send notification in Notification Center
	NotificationC->PresentNotification(Notification.get());
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::ActionCancelScheduledExecute(TObject *Sender)
{
	// Providing the fact that you already have a MyNotification previously issued
	NotificationC->CancelNotification("MyNotification");
}
// ---------------------------------------------------------------------------
void __fastcall TNotificationsForm::ActionCancelAllNotificationsExecute(TObject *Sender)
{
	NotificationC->CancelAll();
}
// ---------------------------------------------------------------------------

File: Object Pascal/Mobile Snippets/Notifications/SendCancelNotification/uMain.pas
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.
//
// This software is the copyrighted property of Embarcadero Technologies, Inc.
// ("Embarcadero") and its licensors. You may only use this software if you
// are an authorized licensee of Delphi, C++Builder or RAD Studio
// (the "Embarcadero Products").  This software is subject to Embarcadero's
// standard software license and support agreement that accompanied your
// purchase of the Embarcadero Products and is considered a Redistributable,
// as such term is defined thereunder. Your use of this software constitutes
// your acknowledgement of your agreement to the foregoing software license
// and support agreement.
//---------------------------------------------------------------------------
unit uMain;

interface

uses
  System.Actions, System.Classes, System.Notification,
  FMX.ActnList, FMX.Controls, FMX.Controls.Presentation, FMX.Forms, FMX.Memo, 
  FMX.Memo.Types, FMX.ScrollBox, FMX.StdCtrls, FMX.Types;

type
  TNotificationsForm = class(TForm)
    btnSendScheduledNotification: TButton;
    ToolBar1: TToolBar;
    Label1: TLabel;
    btnSendNotificationImmediately: TButton;
    ToolBar2: TToolBar;
    SpeedButton1: TSpeedButton;
    SpeedButton2: TSpeedButton;
    NotificationC: TNotificationCenter;
    Memo1: TMemo;
    ActionList: TActionList;
    ActionSendScheduledNotification: TAction;
    ActionSendNotificationImmediately: TAction;
    ActionCancelScheduled: TAction;
    ActionCancelAllNotifications: TAction;

    procedure NotificationCPermissionRequestResult(Sender: TObject; const AIsGranted: Boolean);
    procedure NotificationCReceiveLocalNotification(Sender: TObject; ANotification: TNotification);
    procedure ActionListExecute(Action: TBasicAction; var Handled: Boolean);
    procedure ActionSendScheduledNotificationExecute(Sender: TObject);
    procedure ActionSendNotificationImmediatelyExecute(Sender: TObject);
    procedure ActionCancelScheduledExecute(Sender: TObject);
    procedure ActionCancelAllNotificationsExecute(Sender: TObject);
  private
    FPendingAction: TBasicAction;
  end;

var
  NotificationsForm: TNotificationsForm;

implementation

{$R *.fmx}

uses
  System.SysUtils;
  
procedure TNotificationsForm.NotificationCPermissionRequestResult(Sender: TObject; const AIsGranted: Boolean);
begin
  if AIsGranted and (FPendingAction <> nil) then
  begin
    FPendingAction.Execute;
  end;

  FPendingAction := nil;
end;

procedure TNotificationsForm.NotificationCReceiveLocalNotification(Sender: TObject; ANotification: TNotification);
begin
  Memo1.Lines.Add(ANotification.AlertBody);
end;

procedure TNotificationsForm.ActionListExecute(Action: TBasicAction; var Handled: Boolean);
begin
  if NotificationC.AuthorizationStatus <> TAuthorizationStatus.Authorized then
  begin
    Handled := True;
    FPendingAction := Action;

    NotificationC.RequestPermission;
  end;
end;

procedure TNotificationsForm.ActionSendScheduledNotificationExecute(Sender: TObject);
begin
  var Notification := NotificationC.CreateNotification;
  try
    Notification.Name := 'MyNotification';
    Notification.AlertBody := 'Delphi for Mobile is here!';
    { Fired in 10 second }
    Notification.FireDate := Now + EncodeTime(0, 0, 10, 0);

    { Send notification in Notification Center }
    NotificationC.ScheduleNotification(Notification);
  finally
    Notification.Free;
  end;
end;

procedure TNotificationsForm.ActionSendNotificationImmediatelyExecute(Sender: TObject);
begin
  var Notification := NotificationC.CreateNotification;
  try
    Notification.AlertBody := 'Delphi for Mobile is here!';
    Notification.FireDate := Now;

    { Send notification in Notification Center }
    NotificationC.PresentNotification(Notification);
  finally
    Notification.Free;
  end;
end;

procedure TNotificationsForm.ActionCancelScheduledExecute(Sender: TObject);
begin
  { Providing the fact that you already have a MyNotification previously issued }
  NotificationC.CancelNotification('MyNotification');
end;

procedure TNotificationsForm.ActionCancelAllNotificationsExecute(Sender: TObject);
begin
  NotificationC.CancelAll;
end;

end.
